src/benchmark/benchmark.c

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright 2016, General Dynamics C4 Systems
+ *
+ * SPDX-License-Identifier: GPL-2.0-only
+ */
+
+#include <config.h>
+
+#ifdef CONFIG_ENABLE_BENCHMARKS
+
+#include <types.h>
+#include <mode/machine.h>
+#include <benchmark/benchmark.h>
+#include <benchmark/benchmark_utilisation.h>
+
+
+exception_t handle_SysBenchmarkFlushCaches(void)
+{
+#ifdef CONFIG_ARCH_ARM
+    tcb_t *thread = NODE_STATE(ksCurThread);
+    if (getRegister(thread, capRegister)) {
+        arch_clean_invalidate_L1_caches(getRegister(thread, msgInfoRegister));
+    } else {
+        arch_clean_invalidate_caches();
+    }
+#else
+    arch_clean_invalidate_caches();
+#endif
+    return EXCEPTION_NONE;
+}
+
+exception_t handle_SysBenchmarkResetLog(void)
+{
+#ifdef CONFIG_ENABLE_KERNEL_LOG_BUFFER
+    if (ksUserLogBuffer == 0) {
+        userError("A user-level buffer has to be set before resetting benchmark.\
+                Use seL4_BenchmarkSetLogBuffer\n");
+        setRegister(NODE_STATE(ksCurThread), capRegister, seL4_IllegalOperation);
+        return EXCEPTION_SYSCALL_ERROR;
+    }
+
+    ksLogIndex = 0;
+#endif /* CONFIG_ENABLE_KERNEL_LOG_BUFFER */
+
+#ifdef CONFIG_BENCHMARK_TRACK_UTILISATION
+    NODE_STATE(benchmark_log_utilisation_enabled) = true;
+    benchmark_track_reset_utilisation(NODE_STATE(ksIdleThread));
+    NODE_STATE(ksCurThread)->benchmark.schedule_start_time = ksEnter;
+    NODE_STATE(ksCurThread)->benchmark.number_schedules++;
+    NODE_STATE(benchmark_start_time) = ksEnter;
+    NODE_STATE(benchmark_kernel_time) = 0;
+    NODE_STATE(benchmark_kernel_number_entries) = 0;
+    NODE_STATE(benchmark_kernel_number_schedules) = 1;
+    benchmark_arch_utilisation_reset();
+#endif /* CONFIG_BENCHMARK_TRACK_UTILISATION */
+
+    setRegister(NODE_STATE(ksCurThread), capRegister, seL4_NoError);
+    return EXCEPTION_NONE;
+}
+
+exception_t handle_SysBenchmarkFinalizeLog(void)
+{
+#ifdef CONFIG_ENABLE_KERNEL_LOG_BUFFER
+    ksLogIndexFinalized = ksLogIndex;
+    setRegister(NODE_STATE(ksCurThread), capRegister, ksLogIndexFinalized);
+#endif /* CONFIG_ENABLE_KERNEL_LOG_BUFFER */
+
+#ifdef CONFIG_BENCHMARK_TRACK_UTILISATION
+    benchmark_utilisation_finalise();
+#endif /* CONFIG_BENCHMARK_TRACK_UTILISATION */
+
+    return EXCEPTION_NONE;
+}
+
+exception_t handle_SysBenchmarkSetLogBuffer(void)
+{
+#ifdef CONFIG_ENABLE_KERNEL_LOG_BUFFER
+    word_t cptr_userFrame = getRegister(NODE_STATE(ksCurThread), capRegister);
+    if (benchmark_arch_map_logBuffer(cptr_userFrame) != EXCEPTION_NONE) {
+        setRegister(NODE_STATE(ksCurThread), capRegister, seL4_IllegalOperation);
+        return EXCEPTION_SYSCALL_ERROR;
+    }
+#endif /* CONFIG_ENABLE_KERNEL_LOG_BUFFER */
+
+    /* ToDo: If CONFIG_ENABLE_KERNEL_LOG_BUFFER is not set it might be better to
```

Comment:
Can you remove this new comment please.  It would be better suited as an issue instead of as a new ToDo in the code.
Also can you try and not add modifications at the same time as moving large chunks of code as it makes it a lot harder to review.

Reply:
Sorry, I accidentally merged this without giving @axel-h a chance to resolve my comment.

Reply:
I think this should turn into a seL4_Fault_UnknownSyscall if it isn't enabled.

Reply:
I will make a new PR an address this.

Reply:
See https://github.com/seL4/seL4/pull/748